### DISCUSSION TITLE
Get spectra by scannr from mgf files

### DISCUSSION THREAD

**Discussion by JannikSchneider12** [levitsky/pyteomics#131] [General]
Hey,

I have the follwing problem:

I have a tsv file that contains the output from a clustering algorithm of mass spectra. It looks like this:

![image](https://github.com/levitsky/pyteomics/assets/72038436/af9c60c0-4955-4f79-b329-83a1b99e25dc)

So it contains the paths to the mgf files, the scannr of the spectrum and the cluster label.

Now I want to add two columns to that, one containing the m/z array and the other the intensity array of the corresponding spectrum.

My approach getting a spectrum based on the mgf file and scannr is this:

![image](https://github.com/levitsky/pyteomics/assets/72038436/033f3f3f-7377-4622-95f4-cff4af2bca7b)

As far as I tested it, it worked.

Now I want to iterate over the dataframe and add the information to the columns based on the information from the spectra that this function returns. But when I loop over it, it always crashed after the 5th spectrum 

![image](https://github.com/levitsky/pyteomics/assets/72038436/a369a19b-fafb-4b8f-8888-b1d58feaf8df)
![image](https://github.com/levitsky/pyteomics/assets/72038436/7d7c0836-c52d-498e-884c-ded5fd28e02a)

I looked at the mgf file and the information is there. Moreover if I test my single function for this particular case, it works and returns the corresponding information:

![image](https://github.com/levitsky/pyteomics/assets/72038436/fcb545da-0f00-42e1-baf8-59966ec4bb71)

Does anyone has an idea what is wrong?

Thank you for your time an help

**Comment by levitsky** [comment]
Hi @JannikSchneider12,

my hunch is that your problem is related to the way iteration works with the MGF reader. It actually keeps its internal state between your calls to `get_spectrum_by_indexmgf_and_scannr`, so each time you call it, you start where you left off the last time. It works fine as long as you try to retrieve spectra in the same order as they come in the file and without repetition, but if not, the spectrum won't be found.

A simple way to fix this would be to call `indexed_mgf.reset()` on the reader before (or after) each lookup.

However, this usage  does not benefit from the indexing at all. Would it be possible to do without a `for` loop and use random access? If `scannr` corresponds to the actual ordinal number of the spectrum in the file, then you can directly use it to retrieve the spectrum from the indexed reader, like so: `indexed_mgf[scannr]` (perhaps you would need to subtract 1 if `scannr` starts with one and not zero). This way you are going to save a lot of time and you won't need to reset the reader.

If things are more complicated and there is no direct correspondence between `scannr` and the position of the spectrum in the file, then it's possible to subclass `IndexedMGF` for your specific header format so that it indexes spectra by `scannr`. Then you will be able to just retrieve them by `scannr` like from a dictionary. Let me know if you would like help with that.

**Comment by JannikSchneider12** [comment]
Hey @levitsky ,

thank you for the answer. I just sorted my dataframe not only by filename but also by scannr such that there are no problems with the internal state any more. 

So it works but your approach is probably much faster. So the scannr is not directly corresponded with the position in the file although it is also ascending. For now I am using the spectrum.get to get the title and check (like in the first code)

**Reply by levitsky** [reply]
I'm glad that you got it working. If you want to try indexing directly by the `scannr` value, you need a regex that would extract it from the title. Then you can just pass it to the constructor like this:

```python
pyteomics.mgf.IndexedMGF(spectrum_path, use_header=True, label=r"TITLE=.*scan=(\d+)\s*")
```

With the right regex, you can directly access spectra by the matching group (in round brackets). In this case, this should be scan number (but as a string, to not confuse it with positional index), e.g. `indexed_mgf['37408']`.

The performance benefit will only be significant if the dataframe does not reference most of the spectra, though. Otherwise you are essentially parsing the whole file in order, anyway.

**Reply by JannikSchneider12** [reply]
Hey, first of all happy new year and thanks again for your help. This worked really well.

now I have a different problem and I am not sure what is wrong with my approach. So I have created a new mgf file that contains data from several mgf files and I want/have to use a different indexing method. So e.g. one spectrum dictionary looks like this:

{'params': {'title': 'E_Aurora25_huMu-deepC18_dia_387ng-frac8_A3095-8.45.45.2 File:"E_Aurora25_huMu-deepC18_dia_387ng-frac8_A3095-8.raw", NativeID:"controllerType=0 controllerNumber=1 scan=45"', 'pepmass': (536.283053791157, 942706.1137085), 'rtinseconds': 311.2403822 second, 'charge': [2], 'cluster': '152666'}, 'm/z array': array([131.9113617, 149.044342 , 167.0550079, 221.0814362, 223.0626984, 227.7548523, 239.0938721, 239.7316589, 281.050354 , 299.0605164, 300.0608521, 344.9752808, 359.0263977, 389.3624268, 415.034668 , 416.0372925]), 'intensity array': array([ 5641.06494141, 26656.40039062, 22306.484375 , 5434.9140625 , 6765.54150391, 9460.77148438, 8514.15722656, 5090.02050781, 31349.51171875, 105793.015625 , 16241.79296875, 14007.88085938, 20350.1484375 , 5070.68896484, 47915.96875 , 5882.41015625]), 'charge array': masked_array(data=[--, --, --, --, --, --, --, --, --, --, --, --, --, --, --, --], mask=[ True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True], fill_value=0, dtype=int64)}

and for a unique identification I need the first part of the title until the second point. So in this case ‚E_Aurora25_huMu-deepC18_dia_387ng-frac8_A3095-8.45‘

I tried lots of things with regex, but I could not figure out a solution that worked. I know this might be not really pyteomics related and more of a problem to solve with regex, but maybe you have an idea.

thanks for your time and help

**Reply by levitsky** [reply]
Happy New Year to you too!
Here is an example regex that matches the beginning part of the title up to the second period: `r'[^.]*\.[^.]*'`.
Incorporating it into the example above, something like this should work:

```python
pyteomics.mgf.IndexedMGF(spectrum_path, label=r'TITLE=([^.]*\.[^.]*)')
```

**Reply by JannikSchneider12** [reply]
Hey, thanks for your help. I integrated this into my function but I have an issue when I am trying to write a new mgf based on my created spectra dictionaries. So this is my overall function:

![image](https://github.com/levitsky/pyteomics/assets/72038436/d46a10d0-e043-414a-a2c2-4b38526b36df)

And this is my function that creates a new dictionary for the consensus spectra that I want to write inside my new mgf:

![image](https://github.com/levitsky/pyteomics/assets/72038436/737af9a5-1796-4073-8ae7-4f15eba7ce0a)

And I am using this to create a dictionary of the same structure as a spectrum dictionary:

![image](https://github.com/levitsky/pyteomics/assets/72038436/2d98b4f8-e597-4149-a9ac-4d91eaa7d17f)

However when I am writing the mgf file with my new spectra, the mgf file seems not to be recognized by pyteomics, so I can not read it in back. The first entries of the generated file look like this:

BEGIN IONS
PEPMASS=371.313463933932
CHARGE=2+
CLUSTER=16284
111.04401397705078 64523.18359375 
129.0543212890625 93954.2578125 
147.06491088867188 34450.0234375 
149.044921875 30665.169921875 
151.0408477783203 6907.77734375 
167.05514526367188 44286.71484375 
222.90182495117188 7383.04248046875 
223.0625 10910.716796875 
266.9978332519531 42917.4296875 
268.0078125 6828.306640625 
268.977783203125 105841.9609375 
281.0489807128906 11172.298828125 
283.0340576171875 5563.38427734375 
284.92626953125 7698.11474609375 
285.0088806152344 1340683.5 
286.00885009765625 163779.109375 
299.0606689453125 80363.4296875 
300.0614929199219 11942.205078125 
301.058349609375 10812.16015625 
303.0186462402344 31006.560546875 
316.88470458984375 6056.78466796875 
339.0372314453125 33483.97265625 
353.7315673828125 6543.59619140625 
355.06884765625 1040294.875 
356.069580078125 185922.0 
356.12506103515625 11931.4619140625 
370.7531433105469 35299.94921875 
371.7376708984375 257638.21875 
372.7460021972656 50547.14453125 
373.0798645019531 32282.732421875 
389.74774169921875 104781.109375 
540.6638793945312 6793.4482421875 
END IONS

BEGIN IONS
PEPMASS=374.685467727324
CHARGE=2+
CLUSTER=18671
110.39486694335938 5210.1181640625 
115.11697387695312 5041.4443359375 
270.9767761230469 5469.931640625 
285.0102844238281 11201.4169921875 
286.0087890625 25316.61328125 
287.0067138671875 25400.419921875 
288.006103515625 39124.34765625 
289.004150390625 9952.525390625 
356.0693359375 55991.8515625 
357.06658935546875 42618.875 
358.0670471191406 56731.8203125 
359.0672302246094 16768.8671875 
558.2704467773438 4741.7509765625 
END IONS

So is the problem here that the title is missing or is it something different? Again thanks for your time and help

**Reply by JannikSchneider12** [reply]
Update: I included a unique title for every spectrum and now it worked

**Reply by levitsky** [reply]
I'm glad you've sorted it out. Without unique titles, IndexedMGF creates an empty index, but you would still be able to read the file with a non-indexed parser, `MGF`. There's recently been a similar issue, #134, so I'll think about adding some kind of a warning for this.
